pkg/manager/member/tiflash_scaler.go

```diff
@@ -50,113 +52,157 @@ func (s *tiflashScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newS
+				//
+				// The tiflash pod starts for a while, but has not synced its status, and then the pod becomes not ready.
+				// Here we wait for 5 resync periods to ensure that the status of this tiflash pod has been synced.
+				// After this period of time, if there is still no information about this tiflash in TidbCluster status,
+				// then we can be sure that this tiflash has never been added to the tidb cluster.
+				// So we can scale in this tiflash pod safely.
+				return false, fmt.Errorf("TiFlash %s/%s is not ready, wait for some resync periods to synced its status", ns, podName)
+			}
+			klog.Infof("Pod %s/%s not ready for more than %v and no store for it, scale in it",
+				ns, podName, 5*s.deps.CLIConfig.ResyncDuration)
 			err = s.updateDeferDeletingPVC(tc, v1alpha1.TiFlashMemberType, ordinal)
 			if err != nil {
-				return err
+				return false, err
 			}
-			setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
-			return nil
+			return true, nil
 		}
+		return false, fmt.Errorf("tiflash %s/%s no store found in cluster", ns, podName)
 	}
 
-	// When store not found in TidbCluster status, there are two situations as follows:
-	// 1. This can happen when TiFlash joins cluster but we haven't synced its status.
-	//    In this situation return error to wait another round for safety.
-	//
-	// 2. This can happen when TiFlash pod has not been successfully registered in the cluster, such as always pending.
-	//    In this situation we should delete this TiFlash pod immediately to avoid blocking the subsequent operations.
-	if !podutil.IsPodReady(pod) {
-		safeTimeDeadline := pod.CreationTimestamp.Add(5 * s.deps.CLIConfig.ResyncDuration)
-		if time.Now().Before(safeTimeDeadline) {
-			// Wait for 5 resync periods to ensure that the following situation does not occur:
-			//
-			// The tiflash pod starts for a while, but has not synced its status, and then the pod becomes not ready.
-			// Here we wait for 5 resync periods to ensure that the status of this tiflash pod has been synced.
-			// After this period of time, if there is still no information about this tiflash in TidbCluster status,
-			// then we can be sure that this tiflash has never been added to the tidb cluster.
-			// So we can scale in this tiflash pod safely.
-			resetReplicas(newSet, oldSet)
-			return fmt.Errorf("TiFlash %s/%s is not ready, wait for some resync periods to synced its status", ns, podName)
-		}
-		klog.Infof("Pod %s/%s not ready for more than %v and no store for it, scale in it",
-			ns, podName, 5*s.deps.CLIConfig.ResyncDuration)
-		err = s.updateDeferDeletingPVC(tc, v1alpha1.TiFlashMemberType, ordinal)
-		if err != nil {
-			return err
+	var (
+		errs                         []error
+		finishedOrdinals             = sets.NewInt32()
+		updateReplicasAndDeleteSlots bool
+	)
+	for _, ordinal := range ordinals {
+		finished, err := scaleInOne(ordinal)
+		if !finished {
+			errs = append(errs, err)
```

Comment:
Should we break here? Otherwise, the smaller ordinals may be finished and when we update the replicas, the bigger ordinal Pods will be deleted for native sts.

Reply:
OK, so this is handled in L297 in `scaler.go`?

Reply:
We should break here, the other ordinals that finished after this failed ordinal are useless.

Reply:
The first question is handled in function `setReplicasAndDeleteSlotsByFinished` in  `scaler.go`, would count the continuous ordinals in strict order.

Didn't break here since we have to call pd client to delete store for all scaled-in stores, and for the first one `scaleInOne` call of a store it would return a `RequeueError`. If we do not break here, we can combine these `RequeueError` in one round.

What I try to do is: do all the scale-in staff that should be done for all pods, and record successfully or not, leave count strict continuous ordinal when we have to update replica and delete-slots.

But we do have to promise that it would be no harm when call `scaleInOne` multi times for a single ordinal, and I believe this is guaranteed. WDYT?

Reply:
OK, got your point, thanks!